### ISSUE TITLE
How to keep model loaded while running detection in a web service?

### ISSUE THREAD

**Issue by chohaku84** [matterport/Mask_RCNN#600]
I know this might be a newbie question, 
is there anyway to keep model loaded while running detection again and again?

For example, there is a API which can feed images into the detection,
the detection function runs in like 200ms but the config and model-loading run in like 2mins.
So we want to keep config and model loaded so it will save much time for the next image detection.

Any help would be greatly appreciated

**Comment by zungam** [comment]
```
model = modellib.MaskRCNN(mode="inference", config=<your_config_object>,
                                  model_dir=<your_log_dir>)
model.load_weights(<your_weight_path>, by_name=True)

#model is now loaded and kept in RAM
#do something with your model
for img in images:
       output = model.detect([img], verbose=0)[0]
```

**Comment by ItsTehStory** [comment]
@zungam I found the answer about an hour ago but thanks for the quick reply! :)
@mtcld check out my answer here for the fix (https://github.com/matterport/Mask_RCNN/issues/588). But in quick you got to do `model._make_predict_function()`
 before you call 
`model.detect([image], verbose=1)`

**Comment by chohaku84** [comment]
thanks @zungam and @ItsTehStory 
Sorry english is not my first language, I think I just misled you about my question.

Let me clear my question as this:
imagine we have a.py for load the model and weights, b.py just for running the detection on images fed by api.
I know the detection can be run with for-loop within one .py file.
But I just cant figure out how to keep a.py running while b.py being called by API repeatedly.

Maybe I should try to import some framework such as Flask.
But I just want to know if there is someway to do it with scratch development.

**Comment by ItsTehStory** [comment]
Great! Happy you got it working!

**Comment by siontist** [comment]
@chohaku84 can you share how you converted this to flask app? thanks

**Comment by chohaku84** [comment]
@siontist here is a simple implement.  hope it will be of some help.

`# Import libraries
import os
import sys
import random
import math
import re
import time
import numpy as np
import tensorflow as tf
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import imghdr
import cv2
import skimage.io
import logging
import argparse

# Import flask libraries
import hashlib
import json
from time import time
from urllib.parse import urlparse
from uuid import uuid4

import requests
from flask import Flask, jsonify, request

# Root directory of the project
ROOT_DIR = os.path.abspath("../")

# Import Mask RCNN
sys.path.append(ROOT_DIR)  # To find local version of the library
from mrcnn import utils
from mrcnn import visualize
from mrcnn.visualize import display_images
import mrcnn.model as modellib
from mrcnn.model import log
from samples.numberplate import numberplate

# Directories
MODEL_DIR = os.path.join(ROOT_DIR, "logs")      #training logs(weights)
IMAGE_DIR = os.path.join(ROOT_DIR, "images")    #images root folder
MASKED_DIR = os.path.join(IMAGE_DIR, "masked")  #masked images folder
UPLOAD_DIR = os.path.join(IMAGE_DIR, "upload")  #uploaed images folder

# Path to pre-trained weights
NUMBERPLATE_WEIGHTS_PATH = "../weights/"

# Logging confg
logging.basicConfig(level=logging.DEBUG, filename="log", filemode="a+",
                format="%(asctime)-15s %(levelname)-8s %(message)s")

############################################################
#  Configurations
#  Inherits from config.py
############################################################

class InferenceConfig(numberplate.NumberPlateConfig):
    # Set batch size to 1 since we'll be running inference on
    # one image at a time. Batch size = GPU_COUNT * IMAGES_PER_GPU
    GPU_COUNT = 1
    IMAGES_PER_GPU = 1
    IMAGE_RESIZE_MODE = "square"
    IMAGE_MIN_DIM = 512
    IMAGE_MAX_DIM = 576

config = InferenceConfig()
config.display()

# Class names
class_names = ['BG', 'numberplate']

# Create model object in inference mode.
model = modellib.MaskRCNN(mode="inference", model_dir=MODEL_DIR, config=config)

# Load pre-trained weights
NUMBERPLATE_WEIGHTS_FILE = os.path.join(NUMBERPLATE_WEIGHTS_PATH, "mask_rcnn_numberplate.h5")
model.load_weights(NUMBERPLATE_WEIGHTS_FILE, by_name=True)

# VERY VERY IMPORTANT to pre load inference
model.keras_model._make_predict_function()

logging.info('Model and weight have been loaded.')

def run_detect(filename):
    base_file_name = os.path.basename(filename)
    saved_file_name = os.path.join(MASKED_DIR, base_file_name)
    logging.info('Loading image: %s', base_file_name)

    # Convert png with alpha channel with shape[2] == 4 into shape[2] ==3 RGB images
    image = skimage.io.imread(filename)
    if len(image.shape) > 2 and image.shape[2] == 4:
        image = cv2.cvtColor(image, cv2.COLOR_BGRA2BGR)
    logging.info('Image shape info: %s', image.shape)

    # Run detection
    results = model.detect([image], verbose=1)
    r= results[0]
    logging.info('Runing model.detect([image], verbose=1).')

    # Just apply mask then save images
    print_img = visualize.apply_mask_instances(image, r['rois'], r['masks'], r['class_ids'], class_names, r['scores'],None,None,None,None,None,[(1.0,1.0,1.0)])
    skimage.io.imsave(saved_file_name,print_img)
    logging.info('Finished apply_mask_instances.')

    return True

# Instantiate the Node
app = Flask(__name__)

@app.route('/applyMask', methods=['GET','POST'])
def apply_mask_get():

    # Validation images extension
    image_type_ok_list = ['jpeg','png','gif','bmp']

    if request.method == 'GET':

        image_name = request.args.get('image', default = '*', type = str)
        full_filename = os.path.join(IMAGE_DIR, image_name)
        image_type = imghdr.what(full_filename)

        if image_type in image_type_ok_list:
            r = run_detect(full_filename)
            response_msg = "Done. Applied mask into 1 image '%s'" % full_filename
            response = {
                'message': response_msg
            }
            return jsonify(response), 200
        else:
            response = {
                'message': "Invalid image type.(Allowed image type: JPEG, PNG, GIF, BMP)"
            }
            return jsonify(response), 400

    if request.method == 'POST':

        if 'file' not in request.files:
            response = {
                'message': "No file uploaded within the POST body."
            }
            return jsonify(response), 400

        uploaded_file = request.files['file']
        full_filename = os.path.join(UPLOAD_DIR, uploaded_file.filename)
        uploaded_file.save(full_filename)

        result = run_detect(full_filename)

        response_msg = "Done. Applied mask into 1 image '%s'" % full_filename
        response = {
            'message': response_msg
        }
        return jsonify(response), 200

if __name__ == '__main__':

    from argparse import ArgumentParser
    parser = ArgumentParser()
    parser.add_argument('-p', '--port', default=5000, type=int, help='port to listen on')
    args = parser.parse_args()
    port = args.port

    app.run(host='0.0.0.0', port=port)`

**Comment by luisito93** [comment]
could you help me how to keep a model of keras loaded in django please

**Comment by ItsTehStory** [comment]
@luisito93 
I've never used Django, but I believe that the concept is similar in Django and in Flask : 
- You'd have to have Django route your requests to your liking (lets say /api for this)
- You'd have to write a myModelStuff.py
- Write a function that runs the detect/training (although I only tried with detect)
When you have those, just put your loading of the model out of the function mentioned earlier, so that at import time (I am not a python dev sorry to say stuff like this ^^') , your model will be loaded.
So in yourDjangoServer.py : 
```python
import myModelStuff
# Couldn't find how django routes with just a quick search, I'll use flask as an exemple
@app.route('/api')
def youNameThisWhatYouWant():
    myModelStuff.functionToRunDetection();
    return "your answer here"
```
I hope this can help you, but I also hope you found your solution during the last 5 days.

**Comment by luisito93** [comment]
Thanks @ItsTehStory

**Comment by Jargon4072** [comment]
using `model.keras_model._make_predict_function()`  before calling model.detect() worked for me.

**Comment by satwikk** [comment]
In case if you have developed a class and want to initialize the model in the constructor, `_make_predict_fucntion()` needs to be called in the constructor. Calling that function elsewhere (another function or in flask app) will not work.

As mentioned by @Jargon4072, @chohaku84 and @ItsTehStory 

```python
...
    def __init__(self, image_byte=None, image_type=None):
        self.config = coco.CocoConfig()
        self.model = modellib.MaskRCNN(mode="inference", model_dir="./", config=self.config)
        self.model.load_weights(COCO_MODEL_PATH, by_name=True)
        self.model.keras_model._make_predict_function()
...
```

**Comment by caihy05** [comment]
@satwikk 感谢你的分享解决了我的问题

**Comment by jainakhil88** [comment]
> @siontist here is a simple implementation. hope it will be of some help.
> 
> https://github.com/chohaku84/np_detect/blob/master/api/flask_app.py

Update link : https://github.com/chohaku84/keras_bundled_models_demo/blob/f0b0335813bcf5743ed2176cb595329bc0cdeb00/examples/flask_api/app.py

This solution does not work with streamlit.
